core/src/worker/workflow/workflow_stream.rs

```diff
@@ -0,0 +1,927 @@
+                reason,
+            }) {
+                EvictionRequestResult::EvictionRequested(Some(attempt))
+                | EvictionRequestResult::EvictionAlreadyRequested(Some(attempt)) => attempt <= 1,
+                _ => false,
+            };
+            if should_report {
+                FailedActivationOutcome::Report(tt, cause, failure)
+            } else {
+                FailedActivationOutcome::NoReport
+            }
+        };
+        self.reply_to_complete(
+            &run_id,
+            ActivationCompleteOutcome::ReportWFTFail(outcome),
+            resp_chan,
+        );
+    }
+
+    fn process_post_activation(&mut self, report: PostActivationMsg) {
+        let run_id = &report.run_id;
+
+        if self
+            .get_activation(run_id)
+            .map(|a| a.has_eviction())
+            .unwrap_or_default()
+        {
+            self.evict_run(run_id);
+        };
+
+        // If we reported to server, we always want to mark it complete.
+        self.complete_wft(run_id, report.reported_wft_to_server);
+        if let Some(rh) = self.runs.get_mut(run_id) {
+            // Delete the activation
+            rh.activation.take();
+            // Attempt to produce the next activation if needed
+            rh.check_more_activations();
+        }
+    }
+
+    fn local_resolution(&mut self, msg: LocalResolutionMsg) {
+        let run_id = msg.run_id;
+        if let Some(rh) = self.runs.get_mut(&run_id) {
+            rh.send_local_resolution(msg.res)
+        } else {
+            // It isn't an explicit error if the machine is missing when a local activity resolves.
+            // This can happen if an activity reports a timeout after we stopped caring about it.
+            debug!(run_id = %run_id,
+                   "Tried to resolve a local activity for a run we are no longer tracking");
+        }
+    }
+
+    /// Request a workflow eviction. This will (eventually, after replay is done) queue up an
+    /// activation to evict the workflow from the lang side. Workflow will not *actually* be evicted
+    /// until lang replies to that activation
+    fn request_eviction(&mut self, info: RequestEvictMsg) -> EvictionRequestResult {
+        let activation_has_eviction = self.activation_has_eviction(&info.run_id);
+        if let Some(rh) = self.runs.get_mut(&info.run_id) {
+            let attempts = rh.wft.as_ref().map(|wt| wt.info.attempt);
+            if !activation_has_eviction && rh.trying_to_evict.is_none() {
+                debug!(run_id=%info.run_id, reason=%info.message, "Eviction requested");
+                rh.trying_to_evict = Some(info);
+                rh.check_more_activations();
+                EvictionRequestResult::EvictionRequested(attempts)
+            } else {
+                EvictionRequestResult::EvictionAlreadyRequested(attempts)
+            }
+        } else {
+            warn!(run_id=%info.run_id, "Eviction requested for unknown run");
```

Comment:
Is this for lang requesting eviction from core?

Reply:
Technically this should never be possible otherwise, right?

Reply:
It could be either lang or internally